Implement declarativeNetRequest.updateEnabledRulesets

b-weinstein opened the pull request:
#### dc8ef58782b8f08276f8731a52ebbdb54d08ac9d
<pre>
Implement declarativeNetRequest.updateEnabledRulesets
<a href="https://bugs.webkit.org/show_bug.cgi?id=265616">https://bugs.webkit.org/show_bug.cgi?id=265616</a>
<a href="https://rdar.apple.com/118940027">rdar://118940027</a>

Reviewed by Timothy Hatcher.

When an extension calls updateEnabledRulesets - the following happens:
1) We validate the arguments in the WebProcess to make sure they are of the correct format
2) We call into the UI process to perform the work
3) The UI process makes sure all of the passed identifiers exist, and if they do, we attempt to perform the update

To do this, we modify the enabled state of the given rulesets, and attempt to load the declarativeNetRequest rules again.

If this succeeds, we write the updated state to disk. If it fails, we revert the changes that were made (and will continue to use
the old DNR rules).

This patch also adds tests for getEnabledRulesets and updateEnabledRulesets.

* Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h: Copied from Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRulesetStateFromStorage): Read the modified state of rulesets from State.plist
and apply it.
(WebKit::WebExtensionContext::saveDeclarativeNetRequestRulesetStateToStorage): Merge the given dictionary with the existing one and
save it to disk.
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers): Make sure all of the ruleset identifiers exist and throw
an error if any don&apos;t.
(WebKit::WebExtensionContext::declarativeNetRequestEnabledRulesetCount): Get the count of enabled rulesets.
(WebKit::WebExtensionContext::declarativeNetRequestToggleRulesets): Set the state of the given rulesets.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateEnabledRulesets): Perform the logic described in steps 3+ earlier in the commit message.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Adopt a few constants.
(WebKit::WebExtension::declarativeNetRequestRulesets): These ruleset objects can be changed, so don&apos;t return const.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call loadDeclarativeNetRequestRulesetStateFromStorage at startup. We don&apos;t want to call this each time that the
loadDeclarativeNetRequestRules is called, because we might be attempting to update the enabled rulesets and in that case we don&apos;t want to consult storage
yet.
(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary): Clear any customized rulesets upon update.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Add a completion handler.
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateEnabledRulesets): Perform argument validation.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDeclarativeNetRequest.h:
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfStaticRulesets const): Use a constant.
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfEnabledRulesets const): Ditto.
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfDynamicAndSessionRules const): Ditto.

Canonical link: <a href="https://commits.webkit.org/271410@main">https://commits.webkit.org/271410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf146d920b9fe38f4a5afa2886324a06d7e9908f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28279 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6922 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29638 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30809 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25762 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28776 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4301 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28547 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5688 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24340 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4956 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25338 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31498 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25891 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31386 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5053 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3234 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29140 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6639 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6778 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5495 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5571 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -82,6 +82,9 @@
 static NSString * const lastLoadedDNRHashStateKey = @"LastLoadedDNRHash";
 
+// FIXME: Share this between here and WebExtensionContextAPIDeclarativeNetRequestCocoa.
+static NSString * const declarativeNetRequestModifiedRulesetStateKey = @"DeclarativeNetRequestRulesetState";
```
Instead of needing to share it, maybe it's better to put:
`[m_state removeObjectForKey:declarativeNetRequestModifiedRulesetStateKey];` into a helper method that is implemented in `WebExtensionContextDeclarativeNetRequsetCocoa.mm`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -47,6 +51,128 @@
+        }
+
+        if (!found)
+            return { std::nullopt, toErrorString(@"declarativeNetRequest.updateEnabledRulesets", nil, @"Failed to apply rules. Invalid ruleset id: %@.", (NSString *)identifier) };
```
`@"declarativeNetRequest.updateEnabledRulesets()"`. Also the `NSString` cast might not be needed here.

b-weinstein replied:
Added the parens, but this cast was needed:

/Users/brian_weinstein/OpenSource/Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:100:156 Cannot pass object of non-trivial type 'WTF::String' through variadic function; call will abort at runtime

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -47,6 +51,128 @@
+                continue;
+
+            ruleset.enabled = newValue;
+            rulesetIdentifiersToEnabledState[(NSString *)ruleset.rulesetID] = @(newValue);
```
The `NSString` cast might not be needed. It is automatic in most cases.

b-weinstein replied:
I think it wasn't automatic with the dictionary literal syntax when I first wrote it, but I changed it to use setObject:forKey: and I was able to get rid of the cast.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -47,6 +51,128 @@
+    }
+
+    if (declarativeNetRequestEnabledRulesetCount() - rulesetIdentifiersToDisable.size() + rulesetIdentifiersToEnable.size() > webExtensionDeclarativeNetRequestMaximumNumberOfEnabledRulesets) {
+        completionHandler(toErrorString(@"declarativeNetRequest.updateEnabledRulesets", nil, @"The number of enabled static rulesets exceeds the limit. Only %lu rulesets can be enabled at once.", webExtensionDeclarativeNetRequestMaximumNumberOfEnabledRulesets));
```
`@"declarativeNetRequest.updateEnabledRulesets()"`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -47,6 +51,128 @@
+        declarativeNetRequestToggleRulesets(rulesetIdentifiersToDisable, true, rulesetIdentifiersToEnabledState.get());
+        declarativeNetRequestToggleRulesets(rulesetIdentifiersToEnable, false, rulesetIdentifiersToEnabledState.get());
+
+        completionHandler(toErrorString(@"declarativeNetRequest.updateEnabledRulesets", nil, @"Failed to apply rules."));
```
`@"declarativeNetRequest.updateEnabledRulesets()"`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
```diff
@@ -30,20 +30,43 @@
+        enableRulesetsKey: @[ NSString.class ],
+    };
+
+    if (!validateDictionary(options, @"options", @[ ], types, outExceptionString))
```
Can pass `nil` instead of `@[ ]`.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -146,6 +146,174 @@
+            ]
+        }
+    };
+    auto *rules = @"[ { \"id\" : 1, \"priority\": 1, \"action\" : { \"type\" : \"block\" }, \"condition\" : { \"urlFilter\" : \"frame\" } } ]";
```
Nit: Add newline between lines.

xeenon commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIDeclarativeNetRequest.mm:
```diff
@@ -146,6 +146,174 @@
+            ]
+        }
+    };
+    auto *rules = @"[ { \"id\" : 1, \"priority\": 1, \"action\" : { \"type\" : \"block\" }, \"condition\" : { \"urlFilter\" : \"frame\" } } ]";
```
Nit: Add newline between lines.

b-weinstein committed:
Implement declarativeNetRequest.updateEnabledRulesets
https://bugs.webkit.org/show_bug.cgi?id=265616
rdar://118940027

Reviewed by Timothy Hatcher.

When an extension calls updateEnabledRulesets - the following happens:
1) We validate the arguments in the WebProcess to make sure they are of the correct format
2) We call into the UI process to perform the work
3) The UI process makes sure all of the passed identifiers exist, and if they do, we attempt to perform the update

To do this, we modify the enabled state of the given rulesets, and attempt to load the declarativeNetRequest rules again.

If this succeeds, we write the updated state to disk. If it fails, we revert the changes that were made (and will continue to use
the old DNR rules).

This patch also adds tests for getEnabledRulesets and updateEnabledRulesets.

* Source/WebKit/Shared/Extensions/WebExtensionDeclarativeNetRequestConstants.h: Copied from Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRulesetStateFromStorage): Read the modified state of rulesets from State.plist
and apply it.
(WebKit::WebExtensionContext::saveDeclarativeNetRequestRulesetStateToStorage): Merge the given dictionary with the existing one and
save it to disk.
(WebKit::WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers): Make sure all of the ruleset identifiers exist and throw
an error if any don't.
(WebKit::WebExtensionContext::declarativeNetRequestEnabledRulesetCount): Get the count of enabled rulesets.
(WebKit::WebExtensionContext::declarativeNetRequestToggleRulesets): Set the state of the given rulesets.
(WebKit::WebExtensionContext::declarativeNetRequestUpdateEnabledRulesets): Perform the logic described in steps 3+ earlier in the commit message.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded): Adopt a few constants.
(WebKit::WebExtension::declarativeNetRequestRulesets): These ruleset objects can be changed, so don't return const.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call loadDeclarativeNetRequestRulesetStateFromStorage at startup. We don't want to call this each time that the
loadDeclarativeNetRequestRules is called, because we might be attempting to update the enabled rulesets and in that case we don't want to consult storage
yet.
(WebKit::WebExtensionContext::queueStartupAndInstallEventsForExtensionIfNecessary): Clear any customized rulesets upon update.
(WebKit::WebExtensionContext::compileDeclarativeNetRequestRules): Add a completion handler.
(WebKit::WebExtensionContext::loadDeclarativeNetRequestRules): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIDeclarativeNetRequestCocoa.mm:
(WebKit::WebExtensionAPIDeclarativeNetRequest::updateEnabledRulesets): Perform argument validation.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIDeclarativeNetRequest.h:
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfStaticRulesets const): Use a constant.
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfEnabledRulesets const): Ditto.
(WebKit::WebExtensionAPIDeclarativeNetRequest::maxNumberOfDynamicAndSessionRules const): Ditto.

Canonical link: https://commits.webkit.org/271410@main
